propagation.go

```diff
@@ -41,71 +45,115 @@ var (
 type BuiltinFormat byte
 
 const (
-	// SplitBinary encodes the Span in a SplitBinaryCarrier instance.
+	// Binary encodes the Span for propagation as opaque binary data.
+	//
+	// For Tracer.Inject(): the carrier must be an `io.Writer`.
 	//
-	// The `carrier` for injection and extraction must be a
-	// `*SplitBinaryCarrier` instance.
-	SplitBinary BuiltinFormat = iota
+	// For Tracer.Join(): the carrier must be an `io.Reader`.
+	Binary BuiltinFormat = iota
 
-	// SplitText encodes the Span in a SplitTextCarrier instance.
+	// TextMap encodes the Span in a TextMapCarrier instance.
+	//
+	// For Tracer.Inject(): the carrier must be a `TextMapWriter`.
+	//
+	// For Tracer.Join(): the carrier must be a `TextMapReader`.
+	//
+	// See HTTPHeaderTextMapCarrier for an implementation of both TextMapWriter
+	// and TextMapReader that defers to an http.Header instance for storage.
+	// For example, Inject():
+	//
+	//    carrier := HTTPHeaderTextMapCarrier(httpReq.Header)
+	//    err := span.Tracer().Inject(span, TextMap, carrier)
 	//
-	// The `carrier` for injection and extraction must be a `*SplitTextCarrier`
-	// instance.
+	// Or Join():
+	//
+	//    carrier := HTTPHeaderTextMapCarrier(httpReq.Header)
+	//    span, err := tracer.Join("opName", TextMap, carrier)
+	//
+	TextMap
+
+	// SplitBinary is DEPRECATED
+	SplitBinary
+	// SplitText is DEPRECATED
 	SplitText
+)
+
+// TextMapWriter is the Inject() carrier for the TextMap builtin format. With
+// it, the caller can encode a Span for propagation as entries in a multimap of
+// unicode strings.
+type TextMapWriter interface {
+	// Add a key:value pair to the carrier. Multiple values may be added for a
+	// single (repeated) key.
+	Add(key, val string)
+}
 
-	// GoHTTPHeader encodes the Span into a Go http.Header instance (both the
-	// tracer state and any baggage).
+// TextMapReader is the Join() carrier for the TextMap builtin format. With it,
+// the caller can decode a propagated Span as entries in a multimap of unicode
+// strings.
+type TextMapReader interface {
+	// ReadAll returns TextMap contents via repeated calls to the `handler`
+	// function. If any call to `handler` returns a non-nil error, ReadAll
+	// terminates and returns that error.
 	//
-	// The `carrier` for both injection and extraction must be an http.Header
-	// instance.
-	GoHTTPHeader
-)
+	// NOTE: A single `key` may appear in multiple calls to `handler` for a
+	// single `ReadAllEntries` invocation.
```

Comment:
I am not sure how much help adding `HTTPHeaders` format would provide. The key restrictions imposed by http actually spill out into baggage API, not so much the carriers. And if we did special case HTTP headers, we'd still need a text map for other RPCs (grpc, finagle, etc.)

Strictly speaking, we can even remove the case insensitive restriction and allow any strings for baggage keys, we'd just need tracers to encode the keys before injecting into http. We can provide a guide to implementers how to do that, e.g. assume that Canonical-Mixed-Case-Keys go in unchanged, and all other case combinations are url-encoded. The worst case is people won't be able to read the headers in tcpdump if they don't use Mixed-Case (and that's assuming tracer even keeps baggage in individual headers instead of a single base64 header).